### PR TITLE
Updates for CADET-Python-Simulator

### PR DESCRIPTION
**Pull request by schmoelder**
This PR contains some modifications of CADET-Process to run the [CADET-Python-Simulator](https://jugit.fz-juelich.de/IBG-1/ModSim/cadet/cadet-python-simulator).

Since the CADET-Python-Simulator is still very much experimental, let's see if / when we merge these changes to CADET-Process.

### CODE REVIEW

File: CADETProcess/dynamicEvents/section.py
Comment (daklauss) [inline CADETProcess/dynamicEvents/section.py:1]: Shouldn't it be something like:
```
def __call___(self):
       return value
```
if you want to make it callable? Or what is the point here?

**Review by daklauss — CHANGES_REQUESTED** [review]
See comments. Otherwise a test of densities for coverage like for the ```molecular_weight``` as in
https://github.com/fau-advanced-separations/CADET-Process/blob/dev/tests/test_components.py 
would be a good thing.

File: CADETProcess/processModel/componentSystem.py [hunk 1]
```diff
@@ -136,6 +149,11 @@ def molecular_weight(self):
         """list of float or None: The molecular weights of the subspecies."""
         return [spec.molecular_weight for spec in self.molecular_weight]
 
+    @property
+    def density(self):
+        """list of float or None: The density of the subspecies."""
+        return [spec.density for spec in self.density]
+
```
[the remark this answers is not in the document]
Comment (schmoelder) [reply CADETProcess/processModel/componentSystem.py:156]: Good catch, thanks! :)

File: CADETProcess/dynamicEvents/section.py
Comment (schmoelder) [reply CADETProcess/dynamicEvents/section.py:1]: That's one way to do it but I think this has the same effect. You could try it out for yourself. ;-)

File: CADETProcess/processModel/componentSystem.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (schmoelder) [reply CADETProcess/processModel/componentSystem.py:156]: Do you mind fixing this and adding a test?

[the remark this answers is not in the document]
Comment (daklauss) [reply CADETProcess/processModel/componentSystem.py:156]: I can do

File: CADETProcess/dynamicEvents/section.py
Comment (daklauss) [reply CADETProcess/dynamicEvents/section.py:1]: Ah got it, value is a method. This actually does make sense. I tried it with simple attributes wich cannot work and was confused, thanks.

**Comment by daklauss** [comment]
Oh ... i guess i shouldn't have amended my changes directly to your add densities commit until it is ready to merge ....

**Review by daklauss — APPROVED** [review]

File: CADETProcess/processModel/componentSystem.py [hunk 2]
```diff
@@ -136,6 +151,11 @@ def molecular_weight(self):
         """list of float or None: The molecular weights of the subspecies."""
         return [spec.molecular_weight for spec in self.species]
 
+    @property
+    def density(self):
+        """list of float or None: The density of the subspecies."""
```
Comment (schmoelder) [inline CADETProcess/processModel/componentSystem.py:156]: Can this really return None?

**Comment by schmoelder** [comment]
Thanks Daniel, (almost) looks good to me, just some minor details.

Also, without wanting to overload this PR, what do you think about adding type annotations to entire module? Btw, this is something where Tools like Copilot / ChatGPT can really do the bulk of the work.

**Comment by daklauss** [comment]
Added type annotations with gpt. Fixed some minor ai stuf. Review needed @schmoelder. Ready to merge afterwards.

File: CADETProcess/processModel/componentSystem.py [hunk 3]
```diff
@@ -50,17 +52,25 @@ class Component(Structure):
         Charge of component (including species).
     molecular_weight : list
         Molecular weight of component (including species).
+    density : list
```
Comment (schmoelder) [inline CADETProcess/processModel/componentSystem.py:55]: Hier bitte auch type annotations.

File: CADETProcess/processModel/componentSystem.py [hunk 4]
```diff
@@ -346,7 +389,7 @@ def names(self):
         return names
 
     @property
-    def species(self):
+    def species(self) -> List[str]:
```
Comment (schmoelder) [inline CADETProcess/processModel/componentSystem.py:392]: List -> list (seit Python 3.10) (dann muss man nicht mehr von typing importieren

File: CADETProcess/processModel/componentSystem.py [hunk 5]
```diff
@@ -323,7 +366,7 @@ def indices(self):
         return Dict(indices)
 
     @property
-    def species_indices(self):
+    def species_indices(self) -> TypingDict[str, int]:
```
Comment (schmoelder) [inline CADETProcess/processModel/componentSystem.py:369]: Ich glaube, hier kannst du einfach dict[str, int] machen.

File: CADETProcess/processModel/componentSystem.py [hunk 6]
```diff
@@ -323,7 +366,7 @@ def indices(self):
         return Dict(indices)
 
     @property
-    def species_indices(self):
+    def species_indices(self) -> dict[str, int]:
         """dict: Indices for each species."""
```
Comment (schmoelder) [inline CADETProcess/processModel/componentSystem.py:370]: please also add dict[str, int] here (and in the other docstrings)

File: CADETProcess/processModel/componentSystem.py [hunk 7]
```diff
@@ -222,51 +261,59 @@ def __init__(
...
+            ) -> NoReturn:
         """
         Add a component to the system.
 
         Parameters
         ----------
-        component : {str, Component}
-            The class of the component to be added.
+        component : str | Component
+            The class or name of the component to be added.
```
Comment (schmoelder) [inline CADETProcess/processModel/componentSystem.py:316]: The component instance or name of the component to be added.

**Pull request merged**